src/main/user-impl/java/com/mysql/cj/jdbc/ha/ca/ClusterAwareUtils.java

```diff
@@ -62,4 +62,14 @@ public static HostInfo copyWithAdditionalProps(HostInfo baseHostInfo, Map<String
         mergedProps.putAll(additionalProps);
         return new HostInfo(urlContainer, baseHostInfo.getHost(), baseHostInfo.getPort(), baseHostInfo.getUser(), baseHostInfo.getPassword(), mergedProps);
     }
+
+    public static HostInfo copyWithAdditionalProps(HostInfo baseHostInfo, ConnectionUrl connectionUrl) {
+        if (baseHostInfo == null) {
+            return connectionUrl.getMainHost();
+        }
+
+        return new HostInfo(connectionUrl, baseHostInfo.getHost(), baseHostInfo.getPort(),
+            connectionUrl.getMainHost().getUser(), connectionUrl.getMainHost().getPassword(),
+            connectionUrl.getMainHost().getHostProperties());
```

Comment:
`connectionUrl.getMainHost()` is called three times. Please save it as a variable.